### ISSUE TITLE
Use Nexus (sonatype.org) to cache the download repo and cache on the local repo

### ISSUE THREAD

**Issue by praveensameneni** [opensearch-project/opensearch-build#3956]
### Is your feature request related to a problem? Please describe

Currently, the build system uses mirror sites for downloading the dependencies and we observed an issue with the mirror site being down during OpenSearch 2.9 release. To mitigate the issue for future releases, please consider using  Nexus ([sonatype.org](http://sonatype.org/)) to cache the download repo and cache on the local repo.

### Describe the solution you'd like

Please consider using  Nexus ([sonatype.org](http://sonatype.org/)) to cache the download repo and cache on the local repo.

**Comment by gaiksaya** [comment]
[Triage] Hi @praveensameneni,

Can you add more details on what kind of dependencies we want to cache? 

The direct dependencies are published to maven local and are used directly by OpenSearch and its components. For external dependencies, we might need to create a support ticket with nexus to configure those proxy URLs and accordingly create a campaign to add that URL in all components' build.gradle.

Let me know if I am getting it right or not.
Thanks!

**Comment by praveensameneni** [comment]
@gaiksaya , During the 2.9 release, we experienced an outage on one of the mirror sites hosting distribution of Node and teams were stuck waiting for the site to come back online.
@joshuarrrr , @ashwin-pc , Would you like to chime in here?

**Comment by ashwin-pc** [comment]
I forget which workflow failed that caused it because the github action we use to download and install node does not mention any mirror. https://github.com/actions/setup-node/tree/v3/. Maybe i'm looking at the wrong workflow so looping in @kavilla since he has more context here.

**Comment by bbarani** [comment]
@kavilla @ashwin-pc @AMoo-Miki We would like to know if it really make sense to cache these dependencies using local repo implementation. Do you see significant benefits in doing it? We don't want to create a complex solution for a transient problem. Can you please provide your inputs?

**Comment by kavilla** [comment]
Yeah we had a mirror for downloading and distributing a Node executable with OpenSearch Dashboards.

We originally switched to the mirror because the traffic to the original download was too much and it was faster to switch the mirror. But then they stopped hosting the mirror and proceeded to now download any copy of Node.

IMO, this would be great because instead of every build for each platform/arch we download a ~20 MB file to bundle with the release. So unless the build system has some caching mechanism built in already then the amount of time saved could really add up. I am also currently seeing check sum failures for the build coming from these downloads but that is also a transient issue.

**Comment by bbarani** [comment]
> Yeah we had a mirror for downloading and distributing a Node executable with OpenSearch Dashboards.
> 
> We originally switched to the mirror because the traffic to the original download was too much and it was faster to switch the mirror. But then they stopped hosting the mirror and proceeded to now download any copy of Node.
> 
> IMO, this would be great because instead of every build for each platform/arch we download a ~20 MB file to bundle with the release. So unless the build system has some caching mechanism built in already then the amount of time saved could really add up. I am also currently seeing check sum failures for the build coming from these downloads but that is also a transient issue.

@kavilla Thanks for your inputs. Based on your comments, it looks like the issue that we faced during 2.9.0 was already remediated by moving out of using mirror site and caching it to local would be a "nice to have" feature and not a "must have" feature at this point in time. Do you agree?

**Comment by manasvinibs** [comment]
@bbarani Another [issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5413) we are seeing related to build failures while downloading node executable for v14.20.3 Darwin + arm64 [issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5413) is due to 404 error on access to https://nodejs.org/dist/v14.21.3/node-v14.21.3-darwin-arm64.tar.gz  If we continue to support Darwin arm64 for node v14, this might need to build and cached.

**Comment by prudhvigodithi** [comment]
Hey @manasvinibs there is an open issue related to `node-v14.21.3-darwin-arm64.tar.gz` please check https://github.com/nvm-sh/nvm/issues/3148. Looks like this `14 darwin-arm64 version` is not available to even download and cache.

Ideally we should cache the artifacts for building internet restricted applications, where the build process cannot reach internet to fetch dependencies when running a given build task, but this is not the case for OpenSearch. As @bbarani mentioned its good to have as fallback but my point is its not recommended for Open Source applications.

**Comment by peterzhuamazon** [comment]
Closing as we dont need this anymore, mirror already online.
